<a href="https://colab.research.google.com/github/JustmeLu/Aprendizaje-Automatico/blob/main/2-Ejercicios/01-Mi-segundo-modelo-Plantilla.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/andres-merino/AprendizajeAutomaticoInicial-05-N0105/blob/main/2-Ejercicios/01-Mi-segundo-modelo-Plantilla.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<table style="border: none; border-collapse: collapse;">
    <tr>
        <td style="width: 20%; vertical-align: middle; padding-right: 10px;">
            <img src="https://i.imgur.com/nt7hloA.png" width="100">
        </td>
        <td style="width: 2px; text-align: center;">
            <font color="#0030A1" size="7">|</font><br>
            <font color="#0030A1" size="7">|</font>
        </td>
        <td>
            <p style="font-variant: small-caps;"><font color="#0030A1" size="5">
                <b>Facultad de Ciencias Exactas, Naturales y Ambientales</b>
            </font> </p>
            <p style="font-variant: small-caps;"><font color="#0030A1" size="4">
                Aprendizaje Automático Inicial &bull; Mi segundo modelo
            </font></p>
            <p style="font-style: oblique;"><font color="#0030A1" size="3">
                Nombre y Apellido &bull; Fecha
            </font></p>
        </td>  
    </tr>
</table>

---
## <font color='264CC7'> Introducción </font>

Este cuaderno está diseñado practicar en el entrenamiento y evaluación de un modelo de regresión lineal utilizando datos reales. Los objetivos principales son:

- Entrenar un modelo de regresión lineal.
- Evaluar el ajuste del modelo.


Del conjunto de datos disponible [aquí](https://raw.githubusercontent.com/mGalarnyk/Tutorial_Data/master/King_County/kingCountyHouseData.csv), se desea predecir la variable `price` en función de las variables `bedrooms`, `bathrooms`, `sqft_living`, `sqft_lot`, `floors`.

In [23]:
# Paquetes necesarios
import pandas as pd
from sklearn.preprocessing import StandardScaler, MinMaxScaler, KBinsDiscretizer, OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


---
## <font color='264CC7'> Lectura de Datos </font>


Realiza la lectura de los datos y presenta las primeras filas del conjunto de datos.

In [7]:
# Lectura de los datos desde la URL
url = "https://raw.githubusercontent.com/mGalarnyk/Tutorial_Data/master/King_County/kingCountyHouseData.csv"
data = pd.read_csv(url)

# Mostrar las primeras filas del conjunto de datos
print("Primeras 10 filas del dataset:")
data.head(10)

Primeras 10 filas del dataset:


id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0         3       1.00         1180   
1  6414100192  20141209T000000   538000.0         3       2.25         2570   
2  5631500400  20150225T000000   180000.0         2       1.00          770   
3  2487200875  20141209T000000   604000.0         4       3.00         1960   
4  1954400510  20150218T000000   510000.0         3       2.00         1680   
5  7237550310  20140512T000000  1225000.0         4       4.50         5420   
6  1321400060  20140627T000000   257500.0         3       2.25         1715   
7  2008000270  20150115T000000   291850.0         3       1.50         1060   
8  2414600126  20150415T000000   229500.0         3       1.00         1780   
9  3793500160  20150312T000000   323000.0         3       2.50         1890   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   
5    101930     1.0           0     0  ...     11        3890           1530   
6      6819     2.0           0     0  ...      7        1715              0   
7      9711     1.0           0     0  ...      7        1060              0   
8      7470     1.0           0     0  ...      7        1050            730   
9      6560     2.0           0     0  ...      7        1890              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   
5      2001             0    98053  47.6561 -122.005           4760   
6      1995             0    98003  47.3097 -122.327           2238   
7      1963             0    98198  47.4095 -122.315           1650   
8      1960             0    98146  47.5123 -122.337           1780   
9      2003             0    98038  47.3684 -122.031           2390   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  
5      101930  
6        6819  
7        9711  
8        8113  
9        7570  

[10 rows x 21 columns]

In [8]:
url = 'https://raw.githubusercontent.com/mGalarnyk/Tutorial_Data/master/King_County/kingCountyHouseData.csv'
df = pd.read_csv(url)
# Seleccionar las columnas de interés
columns = ['bedrooms','bathrooms','sqft_living','sqft_lot','floors','price']
df = df.loc[:, columns]
df.head(10)

bedrooms  bathrooms  sqft_living  sqft_lot  floors      price
0         3       1.00         1180      5650     1.0   221900.0
1         3       2.25         2570      7242     2.0   538000.0
2         2       1.00          770     10000     1.0   180000.0
3         4       3.00         1960      5000     1.0   604000.0
4         3       2.00         1680      8080     1.0   510000.0
5         4       4.50         5420    101930     1.0  1225000.0
6         3       2.25         1715      6819     2.0   257500.0
7         3       1.50         1060      9711     1.0   291850.0
8         3       1.00         1780      7470     1.0   229500.0
9         3       2.50         1890      6560     2.0   323000.0

---
## <font color='264CC7'> Preprocesamiento de datos </font>

Realiza un escalado de los datos, prueba diferentes técnicas y selecciona la que mejor se ajuste a los datos.

In [9]:
X = df[['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors']]
Y = df['price']

In [10]:
# Normalización por el máximo
x_max = X.max()
X_max_normalized = X / x_max
print("Normalización por el máximo:\n")
display(X_max_normalized.head())

# Normalización Min-Max
min_max_scaler = MinMaxScaler()
X_min_max_normalized = pd.DataFrame(min_max_scaler.fit_transform(X), columns=X.columns)
print("Normalización Min-Max:\n")
display(X_min_max_normalized.head())

# Estandarización
standard_scaler = StandardScaler()
X_standardized = pd.DataFrame(standard_scaler.fit_transform(X), columns=X.columns)
print("Estandarización:\n")
display(X_standardized.head())

Normalización por el máximo:



bedrooms  bathrooms  sqft_living  sqft_lot    floors
0  0.090909    0.12500     0.087149  0.003421  0.285714
1  0.090909    0.28125     0.189808  0.004385  0.571429
2  0.060606    0.12500     0.056869  0.006056  0.285714
3  0.121212    0.37500     0.144756  0.003028  0.285714
4  0.090909    0.25000     0.124077  0.004893  0.285714

Normalización Min-Max:



bedrooms  bathrooms  sqft_living  sqft_lot  floors
0  0.090909    0.12500     0.067170  0.003108     0.0
1  0.090909    0.28125     0.172075  0.004072     0.4
2  0.060606    0.12500     0.036226  0.005743     0.0
3  0.121212    0.37500     0.126038  0.002714     0.0
4  0.090909    0.25000     0.104906  0.004579     0.0

Estandarización:



bedrooms  bathrooms  sqft_living  sqft_lot    floors
0 -0.398737  -1.447464    -0.979835 -0.228321 -0.915427
1 -0.398737   0.175607     0.533634 -0.189885  0.936506
2 -1.473959  -1.447464    -1.426254 -0.123298 -0.915427
3  0.676485   1.149449    -0.130550 -0.244014 -0.915427
4 -0.398737  -0.149007    -0.435422 -0.169653 -0.915427

---
## <font color='264CC7'> Características y etiquetas </font>

Separa las características y las etiquetas y mostrar la dimensión de cada elemento.

In [12]:
# Suponiendo que 'df' es tu DataFrame original estandarizado
# Y la columna 'price' contiene los precios a predecir

# Separar Características (X) y Etiqueta (Y)
features = ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors']
X = data[features]
y = data['price']

# Mostrar dimensiones
print(f"Dimensiones de las Características (X): {X.shape}")
print(f"Dimensiones de las Etiquetas (Y): {Y.shape}")

Dimensiones de las Características (X): (21613, 5)
Dimensiones de las Etiquetas (Y): (21613,)


---
## <font color='264CC7'> Conjunto de entrenamiento y prueba </font>

Divide los datos en dos conjuntos: uno para entrenamiento y otro para prueba y visualiza la dimensión de cada conjunto. Usar `random_state` igual al últmo número de tu cédula.

In [15]:
# Dividir datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

Revisa la distribución de las etiquetas en los conjuntos de entrenamiento y prueba.

In [16]:
# Reviso las dimensiones de los conjuntos de datos
print('Dimensiones de los conjuntos de datos:')
print('X_train:', X_train.shape)
print('y_train:', y_train.shape)
print('X_test:', X_test.shape)
print('y_test:', y_test.shape)

Dimensiones de los conjuntos de datos:
X_train: (17290, 5)
y_train: (17290,)
X_test: (4323, 5)
y_test: (4323,)


---
## <font color='264CC7'> Entrenamiento </font>

Define un modelo de regresión lineal y entrenalo con los datos de entrenamiento.

In [ ]:
# Crear el modelo y ajustarlo
modelo = LinearRegression()
modelo.fit(X_train, y_train)

# Visualizar los coeficientes
print('Ecuación de la recta:')
print('y =', modelo.coef_[0][0], 'x +', modelo.intercept_[0])

In [19]:
modelo = LinearRegression()
modelo.fit(X_train, y_train)

# Visualizar los coeficientes
print('Ecuación de la recta:')
print(f'y = {modelo.intercept_:.2f} + {modelo.coef_[0]:.2f} * bedrooms + {modelo.coef_[1]:.2f} * bathrooms + {modelo.coef_[2]:.2f} * sqft_living + {modelo.coef_[3]:.2f} * sqft_lot + {modelo.coef_[4]:.2f} * floors')

Ecuación de la recta:
y = 71231.97 + -58230.08 * bedrooms + 4992.33 * bathrooms + 318.09 * sqft_living + -0.37 * sqft_lot + 466.55 * floors


---
## <font color='264CC7'> Evaluación </font>

Realiza la predicciones con el conjunto de prueba y muestra los valores en una tabla.

In [21]:
# Predicciones del modelo
y_pred = modelo.predict(X_test)

# Comparar las predicciones con los datos de prueba
df_predicciones = pd.DataFrame({
    'Actual': y_test.values.flatten(),
    'Predicción': y_pred.flatten()
})
display(df_predicciones)

Actual    Predicción
0      297000.0  4.179096e+05
1     1578000.0  1.321662e+06
2      562100.0  4.156510e+05
3      631500.0  3.186865e+05
4      780000.0  8.596693e+05
...         ...           ...
4318   844000.0  8.132561e+05
4319   335500.0  3.033783e+05
4320   369950.0  3.411705e+05
4321   300000.0  2.026461e+05
4322   575950.0  4.254412e+05

[4323 rows x 2 columns]

Calcula el error cuadrático medio.

In [24]:
# Métricas de evaluación
mse = mean_squared_error(y_test, y_pred)

print('Error cuadrático medio:', mse)

Error cuadrático medio: 59742013410.60567


Escribe el error cuadrático medio para cada tipo de escalado:

- Escalado MaxAbsScaler:
- Escalado MinMaxScaler:
- Escalado RobustScaler:
- Escalado StandardScaler:
- Escalado Normalizer:
- Escalado QuantileTransformer:
- Escalado PowerTransformer:

Escribe un párrafo sobre qué hace cada escalado no visto en clase:

- **Escalado ...:** ...

# Task
The final task is to evaluate the performance of different data scaling techniques on a linear regression model. This involves:

1.  **Describing** the functionality of `MaxAbsScaler`, `RobustScaler`, `Normalizer`, `QuantileTransformer`, and `PowerTransformer`.
2.  **Applying** each of the following scalers (`MaxAbsScaler`, `MinMaxScaler`, `RobustScaler`, `StandardScaler`, `Normalizer`, `QuantileTransformer`, `PowerTransformer`) to the training and test sets.
3.  **Training** a linear regression model on the scaled training data and **calculating the Mean Squared Error (MSE)** on the scaled test data for each scaler.
4.  **Summarizing** all calculated MSE results in a table or list.
5.  **Presenting** the descriptions of the specified scalers and the comprehensive summary of MSE results for each scaling technique.

## Describir Escaladores

### Subtask:
Explicar el funcionamiento de los escaladores `MaxAbsScaler`, `RobustScaler`, `Normalizer`, `QuantileTransformer` y `PowerTransformer`.


### MaxAbsScaler

El `MaxAbsScaler` escala cada característica de forma independiente dividiendo cada valor por el valor absoluto máximo de esa característica. Esto da como resultado que los datos escalados estén dentro del rango `[-1, 1]`. Es útil cuando los datos ya están centrados en cero o contienen valores dispersos, ya que no altera la dispersión y evita distorsionar la distribución original, a diferencia de `MinMaxScaler` que sí afecta la dispersión al ajustar los datos en un rango fijo `[0, 1]` o `[-1, 1]` según sea el caso. Es particularmente adecuado para datos donde se sabe que la media es cero y se desea preservar la información sobre la escasez, como en el procesamiento de datos dispersos.

### RobustScaler

El `RobustScaler` elimina la mediana y escala los datos seg\u00fan el rango intercuart\u00edlico (IQR). El IQR es la diferencia entre el percentil 75 (Q3) y el percentil 25 (Q1). Este escalador es \u00fatil en la presencia de _outliers_ (valores at\u00edpicos), ya que utiliza estad\u00edsticos robustos que son menos sensibles a ellos. A diferencia de `StandardScaler`, que utiliza la media y la desviaci\u00f3n est\u00e1ndar (ambos fuertemente influenciados por los _outliers_), `RobustScaler` garantiza que los valores at\u00edpicos tengan un impacto m\u00ednimo en la transformaci\u00f3n. Los datos se escalan para que la mediana de cada caracter\u00edstica sea cero y el IQR sea uno, lo que ayuda a prevenir que los _outliers_ dominen el rango de los datos.

### Normalizer

El `Normalizer` escala cada muestra (fila) de datos de forma individual para que tenga una norma unitaria. Esto significa que la suma de los cuadrados de los valores de cada fila (si se usa la norma L2) ser&#225; 1. A diferencia de otros escaladores que operan sobre las caracter&#237;sticas (columnas), `Normalizer` act&#250;a sobre las muestras, lo que es &#250;til cuando se trabaja con datos donde la direcci&#243;n de los vectores de caracter&#237;sticas es m&#225;s importante que su magnitud absoluta. Se usa com&#250;nmente en algoritmos como el clustering k-means, la miner&#237;a de texto (tf-idf) o en redes neuronales donde la entrada por muestra debe tener una norma espec&#237;fica. Este escalador no elimina la media ni cambia la forma de la distribuci&#243;n, solo ajusta la magnitud de los vectores.

### QuantileTransformer

El `QuantileTransformer` transforma las caracter&#237;sticas para que sigan una distribuci&#243;n espec&#237;fica (uniforme o normal/Gaussiana). Utiliza el rango de cuantiles para mapear los valores de los datos a una nueva distribuci&#243;n. Primero, estima una funci&#243;n de transformaci&#243;n no lineal basada en los cuantiles de los datos. Luego, mapea cada punto de dato a un valor en la distribuci&#243;n de salida (por defecto, una distribuci&#243;n uniforme en el rango [0, 1]). Alternativamente, se puede especificar una distribuci&#243;n normal (`output_distribution='normal'`) para que los datos transformados sigan una distribuci&#243;n Gaussiana. Esta transformaci&#243;n es &#250;til para mitigar el impacto de los _outliers_ y para hacer que los datos sean m&#225;s adecuados para modelos que asumen una distribuci&#243;n normal, como muchos algoritmos lineales. Es una transformaci&#243;n robusta que puede manejar diferentes tipos de distribuciones de datos de entrada.

### PowerTransformer

El `PowerTransformer` es un tipo de escalador que transforma los datos para que sigan una distribuci&#243;n m&#225;s gaussiana (normal), lo que es beneficioso para muchos modelos predictivos que asumen normalidad en los datos de entrada. Esto lo logra aplicando una transformaci&#243;n de potencia, que puede ser la transformaci&#243;n de Box-Cox o la transformaci&#243;n de Yeo-Johnson. Ambas transformaciones son param&#233;tricas y buscan estabilizar la varianza y minimizar la asimetr&#237;a (skewness) en la distribuci&#243;n de los datos. La transformaci&#243;n de Box-Cox solo funciona con datos positivos, mientras que la de Yeo-Johnson puede aplicarse a datos que contengan valores cero o negativos. Esta t&#233;cnica es &#250;til para mejorar el rendimiento de modelos que son sensibles a la distribuci&#243;n de los datos, como la regresi&#243;n lineal, el an&#225;lisis discriminante lineal y los modelos basados en Gaussianas.

## Evaluar MaxAbsScaler

### Subtask:
Aplicar MaxAbsScaler a los conjuntos de entrenamiento y prueba, entrenar un modelo de regresión lineal y calcular el MSE.


**Reasoning**:
I need to apply the MaxAbsScaler to the training and test sets, train a Linear Regression model, and calculate the Mean Squared Error (MSE) to evaluate this scaling method.



In [25]:
from sklearn.preprocessing import MaxAbsScaler

# 1. Inicializar MaxAbsScaler
maxabs_scaler = MaxAbsScaler()

# 2. Ajustar el escalador a X_train y transformar X_train y X_test
X_train_scaled_maxabs = maxabs_scaler.fit_transform(X_train)
X_test_scaled_maxabs = maxabs_scaler.transform(X_test)

# 3. Crear una nueva instancia de LinearRegression
model_maxabs = LinearRegression()

# 4. Entrenar el modelo con los datos escalados
model_maxabs.fit(X_train_scaled_maxabs, y_train)

# 5. Realizar predicciones sobre los datos de prueba escalados
y_pred_maxabs = model_maxabs.predict(X_test_scaled_maxabs)

# 6. Calcular el Error Cuadrático Medio (MSE)
mse_maxabs_scaler = mean_squared_error(y_test, y_pred_maxabs)

print(f"MSE with MaxAbsScaler: {mse_maxabs_scaler}")

MSE with MaxAbsScaler: 59742013410.60424


**Reasoning**:
The previous step successfully applied the MaxAbsScaler and calculated its MSE. Now, I need to evaluate the MinMaxScaler, train a Linear Regression model with the scaled data, and calculate its Mean Squared Error (MSE).



In [26]:
from sklearn.preprocessing import MinMaxScaler

# 1. Inicializar MinMaxScaler
minmax_scaler = MinMaxScaler()

# 2. Ajustar el escalador a X_train y transformar X_train y X_test
X_train_scaled_minmax = minmax_scaler.fit_transform(X_train)
X_test_scaled_minmax = minmax_scaler.transform(X_test)

# 3. Crear una nueva instancia de LinearRegression
model_minmax = LinearRegression()

# 4. Entrenar el modelo con los datos escalados
model_minmax.fit(X_train_scaled_minmax, y_train)

# 5. Realizar predicciones sobre los datos de prueba escalados
y_pred_minmax = model_minmax.predict(X_test_scaled_minmax)

# 6. Calcular el Error Cuadrático Medio (MSE)
mse_minmax_scaler = mean_squared_error(y_test, y_pred_minmax)

print(f"MSE with MinMaxScaler: {mse_minmax_scaler}")

MSE with MinMaxScaler: 59742013410.60427


**Reasoning**:
I need to apply the RobustScaler to the training and test sets, train a Linear Regression model with the scaled data, and calculate its Mean Squared Error (MSE).



In [27]:
from sklearn.preprocessing import RobustScaler

# 1. Inicializar RobustScaler
robust_scaler = RobustScaler()

# 2. Ajustar el escalador a X_train y transformar X_train y X_test
X_train_scaled_robust = robust_scaler.fit_transform(X_train)
X_test_scaled_robust = robust_scaler.transform(X_test)

# 3. Crear una nueva instancia de LinearRegression
model_robust = LinearRegression()

# 4. Entrenar el modelo con los datos escalados
model_robust.fit(X_train_scaled_robust, y_train)

# 5. Realizar predicciones sobre los datos de prueba escalados
y_pred_robust = model_robust.predict(X_test_scaled_robust)

# 6. Calcular el Error Cuadrático Medio (MSE)
mse_robust_scaler = mean_squared_error(y_test, y_pred_robust)

print(f"MSE with RobustScaler: {mse_robust_scaler}")

MSE with RobustScaler: 59742013410.604256


**Reasoning**:
I need to apply the StandardScaler to the training and test sets, train a Linear Regression model with the scaled data, and calculate its Mean Squared Error (MSE).



In [28]:
from sklearn.preprocessing import StandardScaler

# 1. Inicializar StandardScaler
standard_scaler = StandardScaler()

# 2. Ajustar el escalador a X_train y transformar X_train y X_test
X_train_scaled_standard = standard_scaler.fit_transform(X_train)
X_test_scaled_standard = standard_scaler.transform(X_test)

# 3. Crear una nueva instancia de LinearRegression
model_standard = LinearRegression()

# 4. Entrenar el modelo con los datos escalados
model_standard.fit(X_train_scaled_standard, y_train)

# 5. Realizar predicciones sobre los datos de prueba escalados
y_pred_standard = model_standard.predict(X_test_scaled_standard)

# 6. Calcular el Error Cuadrático Medio (MSE)
mse_standard_scaler = mean_squared_error(y_test, y_pred_standard)

print(f"MSE with StandardScaler: {mse_standard_scaler}")

MSE with StandardScaler: 59742013410.604256


**Reasoning**:
I need to apply the Normalizer to the training and test sets, train a Linear Regression model with the scaled data, and calculate its Mean Squared Error (MSE).



In [29]:
from sklearn.preprocessing import Normalizer

# 1. Inicializar Normalizer
normalizer = Normalizer()

# 2. Ajustar el escalador a X_train y transformar X_train y X_test
# Normalizer opera por filas, no se ajusta en el sentido tradicional de fit_transform en las columnas
# Se aplica directamente a los datos
X_train_scaled_normalizer = normalizer.fit_transform(X_train)
X_test_scaled_normalizer = normalizer.transform(X_test)

# 3. Crear una nueva instancia de LinearRegression
model_normalizer = LinearRegression()

# 4. Entrenar el modelo con los datos escalados
model_normalizer.fit(X_train_scaled_normalizer, y_train)

# 5. Realizar predicciones sobre los datos de prueba escalados
y_pred_normalizer = model_normalizer.predict(X_test_scaled_normalizer)

# 6. Calcular el Error Cuadrático Medio (MSE)
mse_normalizer = mean_squared_error(y_test, y_pred_normalizer)

print(f"MSE with Normalizer: {mse_normalizer}")

MSE with Normalizer: 93211533718.4159


**Reasoning**:
I need to apply the QuantileTransformer to the training and test sets, train a Linear Regression model with the scaled data, and calculate its Mean Squared Error (MSE).



In [30]:
from sklearn.preprocessing import QuantileTransformer

# 1. Inicializar QuantileTransformer
# Usamos output_distribution='normal' para que se parezca a una distribución gaussiana
quantile_transformer = QuantileTransformer(output_distribution='normal', random_state=0)

# 2. Ajustar el escalador a X_train y transformar X_train y X_test
X_train_scaled_quantile = quantile_transformer.fit_transform(X_train)
X_test_scaled_quantile = quantile_transformer.transform(X_test)

# 3. Crear una nueva instancia de LinearRegression
model_quantile = LinearRegression()

# 4. Entrenar el modelo con los datos escalados
model_quantile.fit(X_train_scaled_quantile, y_train)

# 5. Realizar predicciones sobre los datos de prueba escalados
y_pred_quantile = model_quantile.predict(X_test_scaled_quantile)

# 6. Calcular el Error Cuadrático Medio (MSE)
mse_quantile_transformer = mean_squared_error(y_test, y_pred_quantile)

print(f"MSE with QuantileTransformer: {mse_quantile_transformer}")

MSE with QuantileTransformer: 71644345515.40945


**Reasoning**:
I need to apply the PowerTransformer to the training and test sets, train a Linear Regression model with the scaled data, and calculate its Mean Squared Error (MSE).



In [31]:
from sklearn.preprocessing import PowerTransformer

# 1. Inicializar PowerTransformer
power_transformer = PowerTransformer()

# 2. Ajustar el escalador a X_train y transformar X_train y X_test
X_train_scaled_power = power_transformer.fit_transform(X_train)
X_test_scaled_power = power_transformer.transform(X_test)

# 3. Crear una nueva instancia de LinearRegression
model_power = LinearRegression()

# 4. Entrenar el modelo con los datos escalados
model_power.fit(X_train_scaled_power, y_train)

# 5. Realizar predicciones sobre los datos de prueba escalados
y_pred_power = model_power.predict(X_test_scaled_power)

# 6. Calcular el Error Cuadrático Medio (MSE)
mse_power_transformer = mean_squared_error(y_test, y_pred_power)

print(f"MSE with PowerTransformer: {mse_power_transformer}")

MSE with PowerTransformer: 71549546403.4412


El error cuadrático medio (MSE) para cada tipo de escalado es el siguiente:

- Escalado MaxAbsScaler: 59742013410.60424
- Escalado MinMaxScaler: 59742013410.60427
- Escalado RobustScaler: 59742013410.604256
- Escalado StandardScaler: 59742013410.604256
- Escalado Normalizer: 93211533718.4159
- Escalado QuantileTransformer: 71644345515.40945
- Escalado PowerTransformer: 71549546403.4412

El error cuadrático medio (MSE) para cada tipo de escalado es el siguiente:

- Escalado MaxAbsScaler: 59742013410.60424
- Escalado MinMaxScaler: 59742013410.60427
- Escalado RobustScaler: 59742013410.604256
- Escalado StandardScaler: 59742013410.604256
- Escalado Normalizer: 93211533718.4159
- Escalado QuantileTransformer: 71644345515.40945
- Escalado PowerTransformer: 71549546403.4412

## Final Task

### Subtask:
Proporcionar las descripciones de los escaladores no vistos y un resumen completo de los resultados del MSE para cada técnica de escalado.


## Summary:

### Q&A

1.  **¿Cuáles son las descripciones de los escaladores no vistos (`MaxAbsScaler`, `RobustScaler`, `Normalizer`, `QuantileTransformer`, y `PowerTransformer`)?**
    *   **MaxAbsScaler**: Escala cada característica de forma independiente dividiendo cada valor por el valor absoluto máximo, resultando en datos en el rango `[-1, 1]`. Es útil para datos centrados en cero o dispersos, preservando la distribución original.
    *   **RobustScaler**: Elimina la mediana y escala los datos según el rango intercuartílico (IQR), haciéndolo robusto a _outliers_. Los datos escalados tienen una mediana de cero y un IQR de uno.
    *   **Normalizer**: Escala cada muestra (fila) de datos individualmente para que tenga una norma unitaria (L2 por defecto). Es útil cuando la dirección de los vectores de características es más importante que su magnitud absoluta.
    *   **QuantileTransformer**: Transforma las características para seguir una distribución específica (uniforme o normal) utilizando cuantiles. Mitiga el impacto de los _outliers_ y adapta los datos a modelos que asumen normalidad.
    *   **PowerTransformer**: Transforma los datos para que sigan una distribución más gaussiana (normal) usando transformaciones de potencia (Box-Cox o Yeo-Johnson). Estabiliza la varianza y minimiza la asimetría.

2.  **¿Cuál es el resumen completo de los resultados del MSE para cada técnica de escalado?**
    *   **MaxAbsScaler**: 59,742,013,410.60
    *   **MinMaxScaler**: 59,742,013,410.60
    *   **RobustScaler**: 59,742,013,410.60
    *   **StandardScaler**: 59,742,013,410.60
    *   **Normalizer**: 93,211,533,718.42
    *   **QuantileTransformer**: 71,644,345,515.41
    *   **PowerTransformer**: 71,549,546,403.44

### Data Analysis Key Findings

*   Los escaladores `MaxAbsScaler`, `MinMaxScaler`, `RobustScaler` y `StandardScaler` obtuvieron resultados de MSE prácticamente idénticos y los más bajos, alrededor de \$5.97 \times 10^{10}\$.
*   El escalador `Normalizer` mostró el rendimiento más bajo con un MSE significativamente mayor de \$9.32 \times 10^{10}\$.
*   `QuantileTransformer` y `PowerTransformer` se desempeñaron de manera intermedia, con MSEs de \$7.16 \times 10^{10}\$ y \$7.15 \times 10^{10}\$ respectivamente, siendo ligeramente mejores que `Normalizer` pero peores que los escaladores lineales.

### Insights or Next Steps

*   Para este conjunto de datos y un modelo de regresión lineal, los escaladores que preservan la forma general de la distribución de los datos o los escalan linealmente (`MaxAbsScaler`, `MinMaxScaler`, `RobustScaler`, `StandardScaler`) son los más efectivos. El `Normalizer` es el menos adecuado, posiblemente debido a que opera por fila y altera la magnitud relativa de las características dentro de cada muestra.
*   Explorar el impacto de los _outliers_ en el rendimiento de los escaladores lineales, ya que `RobustScaler` (diseñado para ser resistente a _outliers_) mostró un rendimiento casi idéntico a otros escaladores como `StandardScaler` y `MinMaxScaler`, lo que podría sugerir que los _outliers_ no son un problema significativo para este conjunto de datos o modelo.
